// eslint-plugin-jss/tests/lib/rules/sort-jss-keys.js
/**
 * @fileoverview Sort keys in JSS
 * @author Lucas L.
 */
'use strict';

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

var rule = require('../../../lib/rules/sort-jss-keys'),
  RuleTester = require('eslint').RuleTester;

var globalHelpers = require('../../../lib/helpers/global');
var transpile = globalHelpers.transpile;
var transpileErrors = globalHelpers.transpileErrors;

//------------------------------------------------------------------------------
// Tests
//------------------------------------------------------------------------------

var ruleTester = new RuleTester();
ruleTester.run('sort-jss-keys', rule, {
  valid: [
    "var obj = { left: 0, position: 'relative' }",
    "var obj = { display: 'block', textAlign: 'center' }",
    "var obj = { backgroundColor: 'red', color: '#fff' }",
    "var obj = { height: 100, width: '100%' }",
    "var obj = { display: 'none', '&:hover': { backgroundColor: 'red', color: 'blue' } }",
    transpile(
      `var obj = { display: 'none', '&:hover': { backgroundColor: 'red', color: 'blue' } }`
    ),
    transpile(`export default {
      noResultsHeader: {
        ...Fonts.Headings.Level4,
      },
    
      noResultsContainer: {
        paddingBottom: 20,
      },
    }`),
    transpile(`
    export default {
      textBox: {
        display: 'none',
      },

      itemInfoSidebar: {
        extend: 'textBox',
        alignItems: 'center',
        margin: 0,
        position: 'absolute',
        right: 0,
        top: 0,
        width: SIDEBAR_WIDTH,
        zIndex: 1,
  
        '& > :last-child': { marginBottom: Margin.BetweenSections.Desktop },
      },

      ['@media (min-width: 120px)']: { 
        textBox: {
          display: 'block',
        }
      },

      ['@media (min-width: 320px)']: { 
        textBox: {
          display: 'none',
        }
      }
    }`)
  ],

  invalid: [
    {
      code: "var obj = { position: 'relative', left: 0 }",
      errors: [
        {
          message:
            "Expected object keys to be in ascending order. 'left' should be before 'position'",
          type: 'Property'
        }
      ]
    },
    {
      code: "var obj = { textAlign: 'center', display: 'block' }",
      errors: [
        {
          message:
            "Expected object keys to be in ascending order. 'display' should be before 'textAlign'",
          type: 'Property'
        }
      ]
    },
    {
      code: "var obj = { color: '#fff', backgroundColor: 'red' }",
      errors: [
        {
          message:
            "Expected object keys to be in ascending order. 'backgroundColor' should be before 'color'",
          type: 'Property'
        }
      ]
    },
    {
      code: "var obj = { width: '100%', height: 100 }",
      errors: [
        {
          message:
            "Expected object keys to be in ascending order. 'height' should be before 'width'",
          type: 'Property'
        }
      ]
    },
    transpileErrors(
      `var obj = { '&:hover': { backgroundColor: 'red', color: 'blue' },  display: 'none' }`,
      [
        {
          message:
            "Expected object keys to be in ascending order. 'display' should be before '&:hover'",
          type: 'Property'
        }
      ]
    ),
    transpileErrors(
      `var obj = { 
        container: { 
          display: 'none', 
          '&:hover': { 
            backgroundColor: 'red', 
            color: 'blue' 
          }, 
          '&:active': { 
            backgroundColor: 'red', 
            color: 'blue' 
          } 
        },

        ['@media (min-width: 420px)']: { 
          textBox: {
            display: 'block',
          }
        },

        ['@media (min-width: 320px)']: { 
          textBox: {
            display: 'none',
          }
        }
      }`,
      [
        {
          message:
            "Expected object keys to be in ascending order. '&:active' should be before '&:hover'",
          type: 'Property'
        }
      ]
    )
  ]
});
